Add prim, ref, and OpInfo for arange

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81826
* __->__ #81734

Per title. 
- See https://github.com/pytorch/pytorch/issues/81959 for discussion on overloading

TODO:
- ~Handle remaining TensorOptions: layout, pin_memory (won't do in this PR)~
- ~Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)~

soulitzer committed:
[WIP] Add ref and OpInfo for arange

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions
- Add sample inputs for floating point and complex numbers
- Why are creation ops not properly registered to decomposition table
- Rest of the error inputs

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers

[ghstack-poisoned]

soulitzer commented on torch/_refs/__init__.py:
```diff
@@ -2985,6 +2986,33 @@ def empty_like(
 
 
+@out_wrapper()
+@register_decomposition(torch.ops.aten.arange)
```
Maybe this shouldn't be registered as a decomposition

soulitzer committed:
Update on "[WIP] Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)

[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)

[ghstack-poisoned]

soulitzer commented:
cc @nkaretnikov

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)

[ghstack-poisoned]

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)

[ghstack-poisoned]

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)

[ghstack-poisoned]

soulitzer commented on aten/src/ATen/native/Itertools.cpp:
```diff
@@ -10,7 +10,7 @@ using namespace at;
 Tensor _triu_mask(int64_t n, int64_t dims, bool diagonal, TensorOptions opt) {
   // get a mask that has value 1 whose indices satisfies i < j < k < ...
   // or i <= j <= k <= ... (depending on diagonal)
-  Tensor range = at::arange(n, opt.dtype(kLong));
```
don't need anymore

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)

[ghstack-poisoned]

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)

[ghstack-poisoned]

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)

[ghstack-poisoned]

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


TODO:
- Properly handle TensorOptions (won't do in this PR)
- Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)

[ghstack-poisoned]

soulitzer commented:
@ngimel - this should be ready for review now!

soulitzer commented on torch/_prims/__init__.py:
```diff
@@ -2474,6 +2476,89 @@ def _amin_nvfuser(
+        dtype = torch.int64
+    else:
+        dtype = torch.get_default_dtype()
+        if all(isinstance(arg, complex) for arg in (start, end, step)):
```
TODO: this should be removed

ngimel reviewed (approved):
Awesome!

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


Per title. 
- See https://github.com/pytorch/pytorch/issues/81959 for discussion on overloading

TODO:
- ~Handle remaining TensorOptions: layout, pin_memory (won't do in this PR)~
- ~Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)~

[ghstack-poisoned]

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


Per title. 
- See https://github.com/pytorch/pytorch/issues/81959 for discussion on overloading

TODO:
- ~Handle remaining TensorOptions: layout, pin_memory (won't do in this PR)~
- ~Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)~

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

eellison commented on test/test_meta.py:
```diff
@@ -405,6 +405,7 @@ def run_meta_crossref(
 meta_function_expected_failures = {
     torch.Tensor.item: {b8, bf16, c128, c64, f16, f32, f64, i16, i32, i64, i8, u8},  # aten::_local_scalar_dense
     torch.Tensor.to_sparse: {b8, bf16, f16, f32, f64, i16, i32, i64, i8, u8},  # aten::to_sparse, aten::to_sparse.sparse_dim
+    torch.arange: {bf16, f16, f32, f64, i16, i32, i64, i8, u8},
```
Why does this introduce `arange` failures ?

soulitzer replied:
Running into the error: `RuntimeError: device type cuda does not match device type of out parameter (meta)` - could it be because arange does not take any tensor inputs?

kit1980 commented:
@pytorchbot revert -m "Broke trunk slow tests https://hud.pytorch.org/pytorch/pytorch/commit/67dc9abbd8bb0456ae5d96d45d7edd219ca9e421" -c nosignal

soulitzer committed:
Update on "Add prim, ref, and OpInfo for arange"


Per title. 
- See https://github.com/pytorch/pytorch/issues/81959 for discussion on overloading

TODO:
- ~Handle remaining TensorOptions: layout, pin_memory (won't do in this PR)~
- ~Add sample inputs for floating point and complex numbers (done for floating point, won't do for complex in this PR)~

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g
